### ISSUE TITLE
devel/release/doc/jobs do not check out feature branch of ros_buildfarm

### ISSUE THREAD

**Issue by jjekircp** [ros-infrastructure/ros_buildfarm#105]
I ran into this a few times when hacking on the buildfarm...I fork the ros_buildfarm repository, check out a feature branch, and run generate_all_jobs.py from that branch.
The maintenance jobs (those that have a Jenkins SCM plugin that checks out ros_buildfarm) pick up my fork and branch properly and run without difficulty.
However, jobs that are generated by one of the *_reconfigure_jobs jobs (i.e. individual doc, sourcedeb, binarydeb, and devel jobs) end up checking out the master branch of my fork during the "clone ros_buildfarm" step.

It seems the logic in the get_repository() function in git.py might not quite be doing the right thing, as every time I fix it manually and then re-run the *_reconfigure_jobs job I see the job config get set back to use the master branch.

Flagging this so other people are aware - I'll try to poke at it if I have a chance.

**Comment by dirk-thomas** [comment]
I am pretty sure that this is a usage pattern which is supposed to work. We have used that many times already. Can you share more information like:
* the setup up your forked repo
* the state of your local working copy
* the commands you run

**Comment by jjekircp** [comment]
I'm trying to run off the `integration` branch on github.com/clearpathrobotics/ros_buildfarm.

Assuming I've already sourced the venv containing the environment as specified in the docs:
```
$ pwd
/home/jjekir/tmp/tool-164-ssh-git-repo/ros_buildfarm

$ echo $PYTHONPATH
/home/jjekir/tmp/tool-164-ssh-git-repo/ros_buildfarm:/opt/ros/indigo/lib/python2.7/dist-packages

$ git status
On branch integration
nothing to commit, working directory clean

$ git remote -v
origin	git@github.com:clearpathrobotics/ros_buildfarm.git (fetch)
origin	git@github.com:clearpathrobotics/ros_buildfarm.git (push)
upstream	https://github.com/ros-infrastructure/ros_buildfarm.git (fetch)
upstream	https://github.com/ros-infrastructure/ros_buildfarm.git (push)

$ ./generate_all_jobs.py http://<server>/<path>/ros_buildfarm_config/raw/master/index.yaml
```

I do get the following warning message (apologies if I simply shouldn't have neglected setting the fallback URL):

```
The git repository url 'https://github.com/clearpathrobotics/ros_buildfarm.git' is different than the fallback url 'https://github.com/ros-infrastructure/ros_buildfarm.git' defined in '/home/jjekir/tmp/tool-164-ssh-git-repo/ros_buildfarm/ros_buildfarm/git.py'
You might want to update the fallback url to ensure that your forked version continues to work correctly when being used outside of a git working copy (e.g. after being installed or from a tarball)
The git repository version 'integration' is different than the Python package version '0.2.1-master'
You might want to update the Python package version to ensure that your forked version continues to work correctly when being used outside of a git working copy (e.g. after being installed or from a tarball)
```

Anyway, then I go run the `Irel_reconfigure-jobs` job. It pulls the right ros_buildfarm:
```
  <scm class="hudson.plugins.git.GitSCM" plugin="git@2.3.4">
    <configVersion>2</configVersion>
    <userRemoteConfigs>
      <hudson.plugins.git.UserRemoteConfig>
        <url>https://github.com/clearpathrobotics/ros_buildfarm.git</url>
      </hudson.plugins.git.UserRemoteConfig>
    </userRemoteConfigs>
    <branches>
      <hudson.plugins.git.BranchSpec>
        <name>integration</name>
      </hudson.plugins.git.BranchSpec>
    </branches>
```

Then I look at the configuration for one of my release jobs, `Isrc_uT__occupancy_grid_utils__ubuntu_trusty__source`

```
    <hudson.tasks.Shell>
      <command>echo "# BEGIN SECTION: Clone ros_buildfarm"
rm -fr ros_buildfarm
git clone --depth 1 -b master https://github.com/clearpathrobotics/ros_buildfarm.git ros_buildfarm
git -C ros_buildfarm log -n 1
rm -fr ros_buildfarm/.git
rm -fr ros_buildfarm/doc
echo "# END SECTION"</command>
    </hudson.tasks.Shell>
```
Please let me know if there is any other information that would be useful.

**Comment by dirk-thomas** [comment]
Can you please either link to the console output of the reconfigure job or post the output here / link to a gist containing it.

**Comment by dirk-thomas** [comment]
From what I can see from the code sourcedeb jobs should be using the same `ros_buildfarm` repo and version as from which they are be generated from: https://github.com/ros-infrastructure/ros_buildfarm/blob/8ebdfad3b3c45042bb749a7f568018d91d440116/ros_buildfarm/release_job.py#L507

The case that your line includes `-b master` indicates that is was generated from the master branch. If it just failed to detect the branch the clone would not use a branch name at all.

Since we currently don't have a working test deployment I can't reproduce the case either.

**Comment by jjekircp** [comment]
I found some information that might be of interest.
I'm going to run a few commit git commands on two different checkouts of ros_buildfarm: one is on my desktop machine where I do development, and the other is in the `/home/jenkins-slave/workspace/Idev_reconfigure-jobs` directory on the master.

My machine:
````
$ git branch -a
  devel_custom_rosdep
* integration
  master
  ssh_git_clone
  remotes/origin/HEAD -> origin/master
  remotes/origin/close_files
  remotes/origin/devel_custom_rosdep
  remotes/origin/doxygen_branch
  remotes/origin/full_branch_name_match
  remotes/origin/integration
  remotes/origin/master
  remotes/origin/more_labels
  remotes/origin/ssh_git_clone
  remotes/origin/use_requests_lib
  remotes/upstream/close_files
  remotes/upstream/doxygen_branch
  remotes/upstream/full_branch_name_match
  remotes/upstream/master
  remotes/upstream/more_labels

$ git rev-parse --abbrev-ref HEAD
integration
$ git rev-parse HEAD
abd14ec2b1409cd8fb8814b6b6e51ae4bcc7024c
````
slave_on_master:
```
# git branch -a
* (detached from abd14ec)
  remotes/origin/close_files
  remotes/origin/devel_custom_rosdep
  remotes/origin/doxygen_branch
  remotes/origin/full_branch_name_match
  remotes/origin/integration
  remotes/origin/master
  remotes/origin/more_labels
  remotes/origin/ssh_git_clone
  remotes/origin/use_requests_lib
# git rev-parse --abbrev-ref HEAD
HEAD
# git rev-parse HEAD
abd14ec2b1409cd8fb8814b6b6e51ae4bcc7024c
```
Digging a bit deeper, it seems that if I check out the branch named `integration`, the `git rev-parse --abbrev-ref HEAD` command gives me the desired output of `integration`. However if I check out the corresponding revision by its hash, the output of that same command is the string `HEAD`. Jenkins seems to be checking out the revision, not the branch.

This doesn't answer why I seem to be running into this issue but not you though.

A full transcript of an instance of `Idev_reconfigure_jobs` is available at the following gist:
https://gist.github.com/jjekircp/c0b56a4895b13f99607b

**Comment by jjekircp** [comment]
Interestingly I made the change below and that seemed to work around it. I poked around and noticed the following:

* In the `devel_job.xml.em` template the git checkout is set to a version of `ros_buildfarm_repository.version or master` and always comes out as `master`, so `ros_buildfarm_repository` has to be either `master` or `None`
* In git.py, the `get_repository()` function calls the `_get_git_repository_version()` function to get the version
* In `_get_git_repository_version()`, there is a call to `find_executable` from common.py. In my case this call was failing, so `_get_git_repository_version` always returned `None`. So this change ensure that git is available inside the container used to run the Irel_reconfigure_jobs job.

```
--- a/ros_buildfarm/templates/devel/devel_create_reconfigure_task.Dockerfile.em
+++ b/ros_buildfarm/templates/devel/devel_create_reconfigure_task.Dockerfile.em
@@ -28,7 +28,7 @@ RUN useradd -u @uid -m buildfarm
 # automatic invalidation once every day
 RUN echo "@today_str"

-RUN python3 -u /tmp/wrapper_scripts/apt-get.py update-and-install -q -y python3-catkin-pkg python3-empy python3-pip python3-rosdistro python3-yaml
+RUN python3 -u /tmp/wrapper_scripts/apt-get.py update-and-install -q -y git python3-catkin-pkg python3-empy python3-pip python3-rosdistro python3-yaml

 RUN pip3 install jenkinsapi
 
 USER buildfarm
```

**Comment by dirk-thomas** [comment]
Thanks for looking into it. It seems that `git` is missing for all the reconfigure jobs. I created #106 to address it based on your above patch. Can you please try and confirm that it solves the problems for you.

**Comment by jjekircp** [comment]
I ended up needing to make a second change actually...after ensuring git is installed, in my particular case the code ended up returning a git SHA as the version string, and then running `git clone -b <version>`, which failed as `git clone` can only accept a branch.

If you take a look at the diff below you will see what I'm trying now: I replaced the `git clone` with a series of commands similar to what Jenkins does (i.e. a separate git init, git fetch, git checkout) and that is working for me right now:

https://github.com/clearpathrobotics/ros_buildfarm/compare/integration...clearpathrobotics:issue-105-wrong-branch

**Comment by dirk-thomas** [comment]
The above line seems to not show a diff for me. Anyway I have updated #106 to continue using a shallow  clone when possible and fallback to a full clone + checkout in case of hashes.

Can you please give it a try and confirm that it fixes your use case?

**Comment by jjekircp** [comment]
Apologies for the delay on this update.

I had to revert 4cea2503300231bfb6c86b1ad84ec73862350dce as when running `generate_doc_independent_job.py` the script kept failing with a `TransientParseError` from empy on the `builder_shell_clone-ros-buildfarm.xml.em` file:

```
TransientParseError in template 'snippet/builder_shell_clone-ros-buildfarm.xml.em': expecting end of complex expression
```

I ran the Irel_reconfigure_jobs job and then randomly picked an Isrc job and it failed because it tried to `git clone -b` a hash:

```
07:19:33 + git clone --depth 1 -b b6b2ae70a2eb642733b9a2fb66c77405d2ca45ec https://github.com/clearpathrobotics/ros_buildfarm.git ros_buildfarm
07:19:33 Cloning into 'ros_buildfarm'...
07:19:33 warning: Could not find remote branch b6b2ae70a2eb642733b9a2fb66c77405d2ca45ec to clone.
07:19:33 fatal: Remote branch b6b2ae70a2eb642733b9a2fb66c77405d2ca45ec not found in upstream origin
07:19:33 Unexpected end of command stream
07:19:33 Build step 'Execute shell' marked build as failure
```

I looked at your diff for #106 and in concept it does appear to be doing the right thing. I'm not sure if there's a syntax error in that snippet or if there's something strange in my ros_buildfarm_config.

**Comment by dirk-thomas** [comment]
I don't see anything wrong with 4cea2503300231bfb6c86b1ad84ec73862350dce but I don't have access to a test farm to try it out myself.

**Comment by jjekircp** [comment]
After adding and removing various pieces of text in the snippet I think I have it:
```
if ros_buildfarm_repository.version and re.match('[0-9a-f]{40}', ros_buildfarm_repository.version):
    # can't create a shallow clone for hashes
```
If the apostrophe is removed from that comment, the `TransientParseError` goes away.
Adding a single apostrophe in either comment in any location causes the error to return.
Adding a pair of apostrophes, however, is fine. My guess is that empy is treating them specially in some way, but a quick glance at the docs doesn't tell me why.

That fixes the syntax error. I haven't yet verified proper functionality though.

**Comment by dirk-thomas** [comment]
Fixed in 9e39ea3af0a93550f775699708e8ecce1b2c9381.

**Comment by jjekircp** [comment]
I put together a testing branch with some changes I need to run my buildfarm successfully + your `reconfigure_requires_git` branch and broke it (with 9e39ea3 applied) with the following error:

```
# BEGIN SECTION: Clone ros_buildfarm
14:07:05 + rm -fr ros_buildfarm
14:07:05 + git clone https://github.com/clearpathrobotics/ros_buildfarm.git ros_buildfarm
14:07:05 Cloning into 'ros_buildfarm'...
14:07:07 + git checkout 1e92bd780440108b7091d80c96737fa2e09339af
14:07:07 fatal: Not a git repository (or any of the parent directories): .git
14:07:07 Build step 'Execute shell' marked build as failure
```

Looks like the `git checkout` needs a `-C ros_buildfarm` to see the working copy and do the checkout. I applied https://github.com/clearpathrobotics/ros_buildfarm/commit/3811a7624430a39ceb9735a8b78bfde26f480e8b and it seems to do the right thing:

```
# BEGIN SECTION: Clone ros_buildfarm
14:09:50 + rm -fr ros_buildfarm
14:09:50 + git clone https://github.com/clearpathrobotics/ros_buildfarm.git ros_buildfarm
14:09:50 Cloning into 'ros_buildfarm'...
14:09:51 + git -C ros_buildfarm checkout 1e92bd780440108b7091d80c96737fa2e09339af
14:09:51 Note: checking out '1e92bd780440108b7091d80c96737fa2e09339af'.
14:09:51 
14:09:51 You are in 'detached HEAD' state. You can look around, make experimental
14:09:51 changes and commit them, and you can discard any commits you make in this
14:09:51 state without impacting any branches by performing another checkout.
14:09:51 
14:09:51 If you want to create a new branch to retain commits you create, you may
14:09:51 do so (now or later) by using -b with the checkout command again. Example:
14:09:51 
14:09:51   git checkout -b new_branch_name
14:09:51 
14:09:51 HEAD is now at 1e92bd7... Merge remote-tracking branch 'upstream/reconfigure_requires_git' into test-upstream-106-new
14:09:51 + git -C ros_buildfarm log -n 1
14:09:51 commit 1e92bd780440108b7091d80c96737fa2e09339af
14:09:51 Merge: c8e034b 9e39ea3
14:09:51 Author: Jonathan Jekir <jjekir@clearpathrobotics.com>
14:09:51 Date:   Fri Oct 30 16:57:29 2015 -0400
14:09:51 
14:09:51     Merge remote-tracking branch 'upstream/reconfigure_requires_git' into test-upstream-106-new
14:09:51     
14:09:51     Conflicts:
14:09:51     	ros_buildfarm/templates/snippet/builder_shell_clone-ros-buildfarm.xml.em
14:09:51 + rm -fr ros_buildfarm/.git
14:09:51 + rm -fr ros_buildfarm/doc
14:09:51 + echo # END SECTION
14:09:51 # END SECTION
```

**Comment by jjekircp** [comment]
FYI I've had this running on my test farm (with the change mentioned in my last comment) for the last few days and it's been working perfectly.

**Comment by dirk-thomas** [comment]
Thank you for confirming: addressed in c39e07847a90dd4c6b357d485a58be6c10454b73.

**Comment by dirk-thomas** [comment]
Merged #106.
